### PR TITLE
[core] Add spdlog for rotation filename format

### PR DESCRIPTION
**Pull request by dentiny**
Ray's filename formatting conflict with spdlog's, making it possible to have to have files interleaved.

https://github.com/ray-project/ray/blob/9a03760fc812e7737cd60d96526b104c1497fecb/python/ray/_private/node.py#L828-L830

A detailed example, it's possible to have
- /tmp/ray/session_latest/gcs_server.1.out created by ray's unique filename rule
- /tmp/ray/session_latest/gcs_server.2.out created by spdlog for the rotation naming rule

I tried to upstream a customized naming function to spdlog, but gets rejected.
Ref: https://github.com/gabime/spdlog/pull/3296

### CODE REVIEW

File: thirdparty/patches/spdlog-rotation-file-format.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+diff --git a/include/spdlog/sinks/rotating_file_sink-inl.h b/include/spdlog/sinks/rotating_file_sink-inl.h
+index cf8b9d5c..4504b866 100644
+--- a/include/spdlog/sinks/rotating_file_sink-inl.h
++++ b/include/spdlog/sinks/rotating_file_sink-inl.h
+@@ -50,7 +50,7 @@ SPDLOG_INLINE rotating_file_sink<Mutex>::rotating_file_sink(
+ }
+
+ // calc filename according to index and file extension if exists.
+-// e.g. calc_filename("logs/mylog.txt, 3) => "logs/mylog.3.txt".
++// e.g. calc_filename("logs/mylog.txt, 3) => "logs/mylog-3.txt".
```
Comment (jjyao) [inline thirdparty/patches/spdlog-rotation-file-format.patch:18]: If we do the patch, let's just do `logs/mylog.txt.1`?

Comment (dentiny) [reply thirdparty/patches/spdlog-rotation-file-format.patch]: Sure, updated.

File: thirdparty/patches/spdlog-rotation-file-format.patch [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+diff --git a/include/spdlog/sinks/rotating_file_sink-inl.h b/include/spdlog/sinks/rotating_file_sink-inl.h
+index cf8b9d5c..0efcc258 100644
+--- a/include/spdlog/sinks/rotating_file_sink-inl.h
++++ b/include/spdlog/sinks/rotating_file_sink-inl.h
+@@ -50,7 +50,7 @@ SPDLOG_INLINE rotating_file_sink<Mutex>::rotating_file_sink(
+ }
+
+ // calc filename according to index and file extension if exists.
+-// e.g. calc_filename("logs/mylog.txt, 3) => "logs/mylog.3.txt".
++// e.g. calc_filename("logs/mylog.txt, 3) => "logs/mylog.txt.1".
```
Comment (jjyao) [inline thirdparty/patches/spdlog-rotation-file-format.patch:18]:
```suggestion
+// e.g. calc_filename("logs/mylog.txt, 3) => "logs/mylog.txt.3".
```

**Review by jjyao — APPROVED** [review]

File: thirdparty/patches/spdlog-rotation-file-format.patch (hunk 2 above)
Comment (dentiny) [reply thirdparty/patches/spdlog-rotation-file-format.patch]: Fixed.

**Comment by dentiny** [comment]
Hi @jjyao could you please help me merge? The CI has passed

**Review by jjyao — APPROVED** [review]

**Pull request merged**